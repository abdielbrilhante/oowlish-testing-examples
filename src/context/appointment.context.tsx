import * as React from 'react';
import { AppointmentState } from '../types';
import { appointmentReducer } from './appointment.reducer';

const initialState: AppointmentState = {
  isLoading: false,
  appointments: null,
  error: null,
};

export const appointmentContext = React.createContext({
  // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-empty-function
  async fetchAppointments(search: string) { },
  state: initialState,
});

export function AppointmentProvider({ children }: { children: React.ReactNode }) {
  const [state, dispatch] = React.useReducer(appointmentReducer, initialState);

  const fetchAppointments = React.useCallback(async (search: string) => {
    dispatch({ type: 'SET_LOADING', payload: false });
    try {
      const response = await fetch(`http://localhost:7171/appointments?search=${search}`);
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      const data = await response.json();
      dispatch({ type: 'SET_APPOINTMENTS', payload: data });
    } catch (error) {
      dispatch({ type: 'SET_ERROR', payload: error });
    }
  }, [dispatch]);

  const contextValue = React.useMemo(() => ({
    fetchAppointments,
    state,
  }), [fetchAppointments, state]);

  return (
    <appointmentContext.Provider value={contextValue} >
      {children}
    </appointmentContext.Provider>
  );
}
