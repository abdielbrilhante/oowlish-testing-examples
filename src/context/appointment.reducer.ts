import { AppointmentReducerAction, AppointmentState } from '../types';

export function appointmentReducer(state: AppointmentState, action: AppointmentReducerAction) {
  if (action.type === 'SET_LOADING') {
    return {
      ...state,
      isLoading: true,
    };
  }

  if (action.type === 'SET_ERROR') {
    return {
      ...state,
      error: action.payload,
      isLoading: false,
    };
  }

  if (action.type === 'SET_APPOINTMENTS') {
    return {
      ...state,
      appointments: action.payload,
      isLoading: false,
    };
  }

  return state;
}
