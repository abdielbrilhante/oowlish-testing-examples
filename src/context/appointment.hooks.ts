import * as React from 'react';
import { omit } from 'lodash-es';
import { Appointment, Customer } from '../types';
import { appointmentContext } from './appointment.context';

export interface CustomerAppointmentSet {
  customer: Customer
  appointments: Omit<Appointment, 'customer'>[]
}

export function useAppointmentList() {
  const { state } = React.useContext(appointmentContext);

  const customers = React.useMemo(() => {
    if (!state.appointments) {
      return [];
    }

    const byCustomerId: Record<string, CustomerAppointmentSet> = {};
    for (const appt of state.appointments) {
      byCustomerId[appt.customer.id] = byCustomerId[appt.customer.id] ?? {
        customer: appt.customer,
        appointments: [],
      };

      byCustomerId[appt.customer.id].appointments.push(omit(appt, 'customer'));
    }

    return Object.values(byCustomerId);
  }, [state.appointments]);

  return { ...state, customers };
}

export function useSearchInput() {
  const timer = React.useRef(0);
  const { fetchAppointments } = React.useContext(appointmentContext);

  return React.useCallback((event: React.ChangeEvent) => {
    clearTimeout(timer.current);
    const search = (event.target as HTMLInputElement).value;
    if (search.length > 2) {
      timer.current = setTimeout(() => {
        fetchAppointments(search);
      }, 300);
    }
  }, [fetchAppointments]);
}
