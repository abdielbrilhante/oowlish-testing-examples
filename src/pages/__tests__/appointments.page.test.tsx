import { act, render } from '@testing-library/react';
import * as React from 'react';
import { AppointmentProvider } from '../../context/appointment.context';
import { AppointmentsPage } from '../appointments.page';
import userEvent from '@testing-library/user-event';
import { mockRequest } from '../../shared/testing/fetch.mock';

// 1 - Render
// 2 - Assertions
// 3 - Mock images import
// 4 - Add necessary context
// 5 - Interactions
// 6 - Re-rendering
// 7 - Mocking the API
// 8 - Using snapshots
// 9 - Extra assertions

jest.mock('../../assets', () => ({
  icons: {
    canine: new URL('http://test-assets.com/dog.png'),
    feline: new URL('http://test-assets.com/cat.png'),
  },
}));

jest.useFakeTimers();

async function rerender() {
  await act(async () => {
    const promise = new Promise((resolve) => setTimeout(resolve, 0));
    jest.runOnlyPendingTimers();
    await promise;
  });
}

const customers = [
  {
    id: '1',
    name: 'John Doe',
    avatar: 'img.png',
    pets: [
      {
        id: '1',
        name: 'Spot',
        sex: 'male',
        breed: 'Labrador',
        species: 'canine',
      },
      {
        id: '2',
        name: 'Lassie',
        sex: 'female',
        breed: 'Rough Collie',
        species: 'canine',
      },
    ],
  },
  {
    id: '2',
    name: 'Richard Doe',
    avatar: 'img.png',
    pets: [
      {
        id: '3',
        name: 'Sphinx',
        sex: 'male',
        breed: 'British Shorthair',
        species: 'feline',
      },
    ],
  },
];

describe('appointment listing page', () => {
  it('allows user to search for appointments based on name', async () => {
    const user = userEvent.setup({ delay: null });

    mockRequest(
      'GET',
      'http://localhost:7171/appointments?search=sam',
      {
        status: 200,
        body: [
          {
            id: '1',
            when: '2022-10-10',
            customer: customers[0],
            pet: customers[0].pets[0],
          },
          {
            id: '2',
            when: '2022-10-30',
            customer: customers[0],
            pet: customers[0].pets[1],
          },
          {
            id: '3',
            when: '2022-10-15',
            customer: customers[1],
            pet: customers[1].pets[0],
          },
        ],
      }
    );

    const { asFragment, getByLabelText } = render(
      <AppointmentProvider>
        <AppointmentsPage />
      </AppointmentProvider>
    );

    await user.type(getByLabelText('Search by customer or pet name'), 'sam');
    await rerender();

    expect(asFragment()).toMatchSnapshot();
  });
});
