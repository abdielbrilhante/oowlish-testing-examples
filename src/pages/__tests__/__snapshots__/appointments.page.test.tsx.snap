// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appointment listing page allows user to search for appointments based on name 1`] = `
<DocumentFragment>
  <div>
    <label
      for="search"
    >
      <div>
        Search by customer or pet name
      </div>
      <input
        id="search"
        placeholder="Type in search..."
        type="text"
      />
    </label>
    <ul
      class="customers"
    >
      <li>
        <div
          class="customer-info"
        >
          <img
            alt="John Doe's avatar"
            src="img.png"
          />
          <div>
            <p>
              John Doe
            </p>
            <p>
              2 pets
            </p>
          </div>
        </div>
        <ul
          class="pets"
        >
          <li>
            <img
              alt="canine icon"
              aria-hidden="true"
              src="http://test-assets.com/dog.png"
            />
            <div
              class="pet-info"
            >
              <div>
                Spot
              </div>
              <div>
                <span
                  class="pet-sex"
                >
                  male
                </span>
                 
                <span
                  class="pet-species"
                >
                  Labrador
                </span>
              </div>
              <div>
                10/10/2022
              </div>
            </div>
          </li>
          <li>
            <img
              alt="canine icon"
              aria-hidden="true"
              src="http://test-assets.com/dog.png"
            />
            <div
              class="pet-info"
            >
              <div>
                Lassie
              </div>
              <div>
                <span
                  class="pet-sex"
                >
                  female
                </span>
                 
                <span
                  class="pet-species"
                >
                  Rough Collie
                </span>
              </div>
              <div>
                10/30/2022
              </div>
            </div>
            <div
              class="followup-tag"
            >
              Follow up on 11/30/2022
            </div>
          </li>
        </ul>
      </li>
      <li>
        <div
          class="customer-info"
        >
          <img
            alt="Richard Doe's avatar"
            src="img.png"
          />
          <div>
            <p>
              Richard Doe
            </p>
            <p>
              1 pet
            </p>
          </div>
        </div>
        <ul
          class="pets"
        >
          <li>
            <img
              alt="feline icon"
              aria-hidden="true"
              src="http://test-assets.com/cat.png"
            />
            <div
              class="pet-info"
            >
              <div>
                Sphinx
              </div>
              <div>
                <span
                  class="pet-sex"
                >
                  male
                </span>
                 
                <span
                  class="pet-species"
                >
                  British Shorthair
                </span>
              </div>
              <div>
                10/15/2022
              </div>
            </div>
            <div
              class="followup-tag"
            >
              Follow up on 11/15/2022
            </div>
          </li>
        </ul>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
